policy_agent/policy_agent.py

```diff
@@ -0,0 +1,354 @@
+    """
+
+    def __init__(self):
+        self.q = Queue.Queue()
+        self.pods = dict()
+        self.services = dict()
+        self.endpoints = dict()
+        self.namespaces = dict()
+
+        self.PodWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "pods"))
+        self.SvcWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "services"))
+        self.NsWatcher = Thread(target=_keep_watch,
+                                args=(self.q, "namespaces"))
+        self.EptsWatcher = Thread(target=_keep_watch,
+                                  args=(self.q, "endpoints"))
+
+        self.PodWatcher.daemon = True
+        self.SvcWatcher.daemon = True
+        self.NsWatcher.daemon = True
+        self.EptsWatcher.daemon = True
+
+    def run(self):
+        """
+        PolicyAgent.run() is called at program init to spawn watch threads and parse their responses 
+        """
+        self.PodWatcher.start()
+        self.SvcWatcher.start()
+        self.NsWatcher.start()
+        self.EptsWatcher.start()
+
+        while True:
+            self.read_responses()
+
+    def read_responses(self):
+        """
+        Read Responses pulls a response off the Queue and processes it
+        If no responses remain in the queue, it will trigger a resync for all resource lists
+        """
+        try:
+            response = self.q.get_nowait()
+            r_json = json.loads(response)
+            r_type = r_json["type"]
+            r_obj = r_json["object"]
+            r_kind = r_obj["kind"]
+
+            _log.info("%s: %s" % (r_type, r_kind))
+
+            if r_kind in ["Pod", "Namespace", "Service"]:
+                self.process_resource(action=r_type, kind=r_kind, target=r_obj)
+            else:
+                _log.info(
+                    "Resource %s not Pod, Service, or Namespace" % r_kind)
+
+        except Queue.Empty:
+            self.resync()
+
+    def process_resource(self, action, kind, target):
+        """
+        Takes a target object and an action and updates internal resource pools
+        :param action: String of ["ADDED", "DELETED", "MODIFIED"] (returned by api watch)
+        :param kind: String of ["Pod", "Namespace", "Service", "Endpoints"]
+        :param target: json dict of resource info
+        """
+        # Determine Resource Pool 
+        if kind == "Pod":
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == "Service":
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == "Endpoints":
+            resource_pool = self.endpoints
+            obj = Endopints(target)
+
+        elif kind == "Namespace":
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        else:
+            _log.error("resource %s not Pod, Service, or Namespace" % kind)
+            sys.exit(1)
```

Comment:
I don't think we want to sys.exit() anywhere in this program - we just want to log an error and return.